<a href="https://colab.research.google.com/github/Spokai3/PhanTichDuLieuVaHocSau/blob/main/Lab8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

***1.Giới thiệu về thư viện NLTK:***

Cài đặt thư viện NLTK:

Vào cửa sổ command line, gõ lệnh pip install NLTK

Import thư viện NLTK và download công cụ NLTK

In [2]:












import nltk
nltk.download_shell()

NLTK Downloader
---------------------------------------------------------------------------
    d) Download   l) List    u) Update   c) Config   h) Help   q) Quit
---------------------------------------------------------------------------
Downloader> l

Packages:
  [ ] abc................. Australian Broadcasting Commission 2006
  [ ] alpino.............. Alpino Dutch Treebank
  [ ] averaged_perceptron_tagger Averaged Perceptron Tagger
  [ ] averaged_perceptron_tagger_ru Averaged Perceptron Tagger (Russian)
  [ ] basque_grammars..... Grammars for Basque
  [ ] biocreative_ppi..... BioCreAtIvE (Critical Assessment of Information
                           Extraction Systems in Biology)
  [ ] bllip_wsj_no_aux.... BLLIP Parser: WSJ Model
  [ ] book_grammars....... Grammars from NLTK Book
  [ ] brown............... Brown Corpus
  [ ] brown_tei........... Brown Corpus (TEI XML Version)
  [ ] cess_cat............ CESS-CAT Treebank
  [ ] cess_esp............ CESS-ESP Treebank
  [ ] chat80.....

tải gói gutenberg

In [3]:
nltk.download('gutenberg')

[nltk_data] Downloading package gutenberg to /root/nltk_data...
[nltk_data]   Unzipping corpora/gutenberg.zip.


True

xem tên các tập tin có trong 'gutenberg' bằng lệnh fileids()

In [4]:
gb=nltk.corpus.gutenberg
print('Gutenberg files : ', gb.fileids())

Gutenberg files :  ['austen-emma.txt', 'austen-persuasion.txt', 'austen-sense.txt', 'bible-kjv.txt', 'blake-poems.txt', 'bryant-stories.txt', 'burgess-busterbrown.txt', 'carroll-alice.txt', 'chesterton-ball.txt', 'chesterton-brown.txt', 'chesterton-thursday.txt', 'edgeworth-parents.txt', 'melville-moby_dick.txt', 'milton-paradise.txt', 'shakespeare-caesar.txt', 'shakespeare-hamlet.txt', 'shakespeare-macbeth.txt', 'whitman-leaves.txt']


truy cập nội dung Macbeth của Shakespeare, sử dụng hàm words(), rồi gán cho một biến

In [5]:
macbeth = nltk.corpus.gutenberg.words('shakespeare-macbeth.txt')

dùng hàm len() để biết độ dài văn bản

In [6]:
len(macbeth)

23140

hiển thị 10 từ đầu tiên của tập tin

In [7]:
macbeth[:10]

['[',
 'The',
 'Tragedie',
 'of',
 'Macbeth',
 'by',
 'William',
 'Shakespeare',
 '1603',
 ']']

trích 5 câu đầu tiên của tập tin (1 câu được kẹp trong cặp ngoặc vuông), ta dùng hàm sent()

In [8]:
nltk.download('punkt')
macbeth_sents = nltk.corpus.gutenberg.sents('shakespeare-macbeth.txt')
macbeth_sents[:5]

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


[['[',
  'The',
  'Tragedie',
  'of',
  'Macbeth',
  'by',
  'William',
  'Shakespeare',
  '1603',
  ']'],
 ['Actus', 'Primus', '.'],
 ['Scoena', 'Prima', '.'],
 ['Thunder', 'and', 'Lightning', '.'],
 ['Enter', 'three', 'Witches', '.']]

***2.Tìm 1 từ với NLTK:***

Tìm từ 'Stage' xuất hiện trong văn bản text

In [9]:
text = nltk.Text(macbeth)
text.concordance('Stage')

Displaying 3 of 3 matches:
nts with Dishes and Seruice ouer the Stage . Then enter Macbeth Macb . If it we
with mans Act , Threatens his bloody Stage : byth ' Clock ' tis Day , And yet d
 struts and frets his houre vpon the Stage , And then is heard no more . It is 


Tìm từ xuất hiện trước và sau từ 'Stage'

In [10]:
text.common_contexts(['Stage'])

the_. bloody_: the_,


Tìm từ tương tự từ 'Stage'

In [11]:
text.similar('Stage')

day time face warre ayre king bleeding man reuolt serieant like
knowledge broyle shew head spring heeles hare thane skie


***3.Phân tích tần số của các từ***

muốn xem 10 từ thông dụng nhất trong văn bản xuất hiện bao nhiêu lần, dùng lệnh most_common()

In [12]:
fd = nltk.FreqDist(macbeth)
fd.most_common(10)

[(',', 1962),
 ('.', 1235),
 ("'", 637),
 ('the', 531),
 (':', 477),
 ('and', 376),
 ('I', 333),
 ('of', 315),
 ('to', 311),
 ('?', 241)]

Stopword là những từ thông dụng, thường ít có ý nghĩa trong quá trình phân tích văn bản và thường cần được loại bỏ.

In [13]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [14]:
sw = set(nltk.corpus.stopwords.words('english'))
print(len(sw))
list(sw)[:10]

179


['where', 's', 'll', "didn't", 'for', 'into', 'only', 'until', 'there', 'too']

có 179 stopword trong từ vựng tiếng Anh. Ta sẽ loại bỏ các từ stopword trong biến macbeth

In [15]:
macbeth_filtered = [w for w in macbeth if w.lower() not in sw]
len(macbeth_filtered)

14946

In [16]:
fd = nltk.FreqDist(macbeth_filtered)
fd.most_common(10)

[(',', 1962),
 ('.', 1235),
 ("'", 637),
 (':', 477),
 ('?', 241),
 ('Macb', 137),
 ('haue', 117),
 ('-', 100),
 ('Enter', 80),
 ('thou', 63)]

loại bỏ các dấu câu theo lệnh

In [17]:
import string
punctuation = set(string.punctuation)
macbeth_filtered2 = [w.lower() for w in macbeth if w.lower() not in sw and w.lower() not in punctuation]
fd = nltk.FreqDist(macbeth_filtered2)
fd.most_common(10)

[('macb', 137),
 ('haue', 122),
 ('thou', 90),
 ('enter', 81),
 ('shall', 68),
 ('macbeth', 62),
 ('vpon', 62),
 ('thee', 61),
 ('macd', 58),
 ('vs', 57)]

***4.Lựa chọn các từ trong văn bản***

Rút trích các từ có độ dài nhất, ví dụ các từ có độ dài lớn hơn 12 ký tự, dùng lệnh:

In [18]:
long_words = [w for w in macbeth if len(w) >12]
sorted(long_words)

['Assassination',
 'Chamberlaines',
 'Distinguishes',
 'Gallowgrosses',
 'Metaphysicall',
 'Northumberland',
 'Voluptuousnesse',
 'commendations',
 'multitudinous',
 'supernaturall',
 'vnaccompanied']

Rút trích các từ có chứa chuỗi 'ious'


In [19]:
ious_words = [w for w in macbeth if 'ious' in w]
ious_words = set(ious_words)
sorted(ious_words)

['Auaricious',
 'Gracious',
 'Industrious',
 'Iudicious',
 'Luxurious',
 'Malicious',
 'Obliuious',
 'Pious',
 'Rebellious',
 'compunctious',
 'furious',
 'gracious',
 'pernicious',
 'pernitious',
 'pious',
 'precious',
 'rebellious',
 'sacrilegious',
 'serious',
 'spacious',
 'tedious']

***5.Bigrams và collocations***

Lọc các bigram sau khi đã loại các stopword và các dấu câu, dùng lệnh sau:

In [20]:
bgrms = nltk.FreqDist(nltk.bigrams(macbeth_filtered2))
bgrms.most_common(15)

[(('enter', 'macbeth'), 16),
 (('exeunt', 'scena'), 15),
 (('thane', 'cawdor'), 13),
 (('knock', 'knock'), 10),
 (('st', 'thou'), 9),
 (('thou', 'art'), 9),
 (('lord', 'macb'), 9),
 (('haue', 'done'), 8),
 (('macb', 'haue'), 8),
 (('good', 'lord'), 8),
 (('let', 'vs'), 7),
 (('enter', 'lady'), 7),
 (('wee', 'l'), 7),
 (('would', 'st'), 6),
 (('macbeth', 'macb'), 6)]

Ngoài bigram ra, còn có trigram, sự kết hợp của 3 từ, ta dùng lệnh trigrams()

In [21]:
tgrms = nltk.FreqDist(nltk.trigrams (macbeth_filtered2))
tgrms.most_common(10)

[(('knock', 'knock', 'knock'), 6),
 (('enter', 'macbeth', 'macb'), 5),
 (('enter', 'three', 'witches'), 4),
 (('exeunt', 'scena', 'secunda'), 4),
 (('good', 'lord', 'macb'), 4),
 (('three', 'witches', '1'), 3),
 (('exeunt', 'scena', 'tertia'), 3),
 (('thunder', 'enter', 'three'), 3),
 (('exeunt', 'scena', 'quarta'), 3),
 (('scena', 'prima', 'enter'), 3)]

***6.Sử dụng văn bản trên mạng***

import thư viện và mở url để đọc file

In [22]:
from urllib import request
url = 'http://www.gutenberg.org/files/2554/2554-0.txt'
response = request.urlopen(url)
raw = response.read().decode('utf8')
raw[:75]

'\ufeffThe Project Gutenberg eBook of Crime and Punishment, by Fyodor Dostoevsky\r'

thay bằng các lệnh sau:

In [23]:
from urllib import request
url = 'http://www.gutenberg.org/files/2554/2554-0.txt'
response = request.urlopen(url)
raw = response.read().decode('utf-8-sig')
raw[:75]

'The Project Gutenberg eBook of Crime and Punishment, by Fyodor Dostoevsky\r\n'

In [24]:
tokens = nltk.word_tokenize(raw)
webtext = nltk.Text(tokens)
webtext[:12]

['The',
 'Project',
 'Gutenberg',
 'eBook',
 'of',
 'Crime',
 'and',
 'Punishment',
 ',',
 'by',
 'Fyodor',
 'Dostoevsky']

***7.Rút trích văn bản từ trang html***

cách rút ttich1 văn bản từ trang html

In [25]:
url = "http://news.bbc.co.uk/2/hi/health/2284783.stm"
html = request.urlopen(url).read().decode('utf8')
html[:120]

'<!doctype html public "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">\r\n<html>\r\n<hea'

dùng thư viện bs4(BeautifulSoup) cung cấp cho bạn các trình phân tích cú pháp phù hợp có thể nhận dạng HTML và trích xuất văn bản

In [26]:
from bs4 import BeautifulSoup
raw = BeautifulSoup(html, 'lxml').get_text()
tokens = nltk.word_tokenize(raw)
text = nltk.Text(tokens)

***8.Phân tích cảm xúc người dùng***

download các movie review dùng lệnh sau:

In [27]:
nltk.download('movie_reviews')

[nltk_data] Downloading package movie_reviews to /root/nltk_data...
[nltk_data]   Unzipping corpora/movie_reviews.zip.


True

In [28]:
import random 
reviews = nltk.corpus.movie_reviews
documents = [(list(reviews.words(fileid)), category)
for category in reviews.categories()
for fileid in reviews.fileids(category)]
random.shuffle(documents)

Xem nội dung review đầu tiên (dòng 0, cột 0)

In [29]:
first_review = ' '.join(documents[0][0])
print(first_review)

i ' m not quite sure what to say about mars attacks ! , which is obviously the work of a deranged genius . when tim burton ' s twisted alien invasion comedy really works , it ' s breathtaking and hilarious in equal measure . and when it doesn ' t work , it ' s just dull . i ' m not even sure it works more often than it doesn ' t , but where it counts -- that is , when this gleefully evil invading force from the red planet gets down to the business of blasting us to kingdom come -- mars attacks ! is brilliant . mars attacks ! is based on a rather unsavory series of trading cards released by topps in the 1950s , and it takes its cues from the same sources as this summer ' s independence day -- old alien invasion flicks , disaster movies , and big - budget special effects extravaganzas . but unlike independence day , which was a painfully middle - of - the - road appeal to the hearts , minds and wallets of america , mars attacks ! has a fully developed and very personal sense of wonder ab

Xem kết quả review đầu tiên (dòng 0, cột 1)

In [30]:
documents[0][1]

'pos'

Ta cần tạo bảng phân phối tần số các từ trong copus, bảng này cần chuyển sang dạng list, ta dùng hàm list()

In [31]:
all_words = nltk.FreqDist(w.lower() for w in reviews.words())
word_features = list(all_words)

Bước tiếp theo là xác dịnh một hàm để tính toán các đặc trưng, tức là những từ đủ quan trọng để thiết lập ý kiến của một review

In [32]:
def document_features(document, word_features):
    document_words = set(document)
    features = {}
    for word in word_features:
        features['{}', format(word)] = (word in document_words)
    return features

Khi định nghĩa hàm document_features(), bạn tạo 1 tập các documents

In [33]:
featuresets = [(document_features(d,word_features), c) for (d,c) in documents]
len(featuresets)

2000

Tạo tập train và test: 1500 dòng đầu dùng cho tập train và 500 dòng cho tập test để đánh giá độ chính xác của mô hình

In [34]:
train_set, test_set = featuresets[1500:], featuresets[:500]
classifier = nltk.NaiveBayesClassifier.train(train_set)

Dùng thuật toán Naive Bayes để phân loại, dùng thư viện NLTK. Sau đó tính toán độ chính xác của thuật toán

In [35]:
train_set, test_set = featuresets[1500:],featuresets[:500]
classifier= nltk.NaiveBayesClassifier.train(train_set)
print(nltk.classify.accuracy(classifier, test_set))

0.724


Chúng ta đã hoàn tất việc phân tích, dưới đây các từ với trong số lớn nhất của các review được đánh giá là positive và negative

In [36]:
classifier.show_most_informative_features(10)

Most Informative Features
          ('{}', 'lame') = True              neg : pos    =     15.2 : 1.0
       ('{}', 'luckily') = True              pos : neg    =     10.0 : 1.0
 ('{}', 'uninteresting') = True              neg : pos    =      8.7 : 1.0
      ('{}', 'villains') = True              neg : pos    =      8.7 : 1.0
    ('{}', 'refreshing') = True              pos : neg    =      8.6 : 1.0
       ('{}', 'crafted') = True              pos : neg    =      8.6 : 1.0
       ('{}', 'focuses') = True              pos : neg    =      7.9 : 1.0
     ('{}', 'initially') = True              pos : neg    =      7.9 : 1.0
      ('{}', 'ordinary') = True              pos : neg    =      7.9 : 1.0
         ('{}', 'holes') = True              neg : pos    =      7.4 : 1.0


***9.Bài tập áp dụng***

1. Viết chương trình Python với thư viện NLTK để liệt kê các tên của copus

In [39]:
nltk.download_shell()

NLTK Downloader
---------------------------------------------------------------------------
    d) Download   l) List    u) Update   c) Config   h) Help   q) Quit
---------------------------------------------------------------------------
Downloader> l

Packages:
  [ ] abc................. Australian Broadcasting Commission 2006
  [ ] alpino.............. Alpino Dutch Treebank
  [ ] averaged_perceptron_tagger Averaged Perceptron Tagger
  [ ] averaged_perceptron_tagger_ru Averaged Perceptron Tagger (Russian)
  [ ] basque_grammars..... Grammars for Basque
  [ ] biocreative_ppi..... BioCreAtIvE (Critical Assessment of Information
                           Extraction Systems in Biology)
  [ ] bllip_wsj_no_aux.... BLLIP Parser: WSJ Model
  [ ] book_grammars....... Grammars from NLTK Book
  [ ] brown............... Brown Corpus
  [ ] brown_tei........... Brown Corpus (TEI XML Version)
  [ ] cess_cat............ CESS-CAT Treebank
  [ ] cess_esp............ CESS-ESP Treebank
  [ ] chat80.....

2.Viết chương trình Python với thư viện NLTK để liệt kê danh sách các stopword bằng các ngôn ngữ khác nhau

In [49]:
nltk.download('stopwords')
sw = nltk.corpus.stopwords.words('english')
list(sw)[:10]

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're"]

In [50]:
sw1 = nltk.corpus.stopwords.words('german')
list(sw1)[:10] 

['aber', 'alle', 'allem', 'allen', 'aller', 'alles', 'als', 'also', 'am', 'an']

In [51]:
sw2 = nltk.corpus.stopwords.words('spanish')
list(sw2)[:10] 

['de', 'la', 'que', 'el', 'en', 'y', 'a', 'los', 'del', 'se']

3.Viết chương trình Python với thư viện NLTK để kiểm tra danh sách các stop word bằng các ngôn ngữ khác nhau

In [52]:
print(len(sw))

179


In [53]:
print(len(sw1))

232


In [54]:
print(len(sw2))

313


4.Viết chương trình Python với thư viện NLTK để loại bỏ các stopword từ một văn bản đã cho

In [60]:
macbeth = nltk.corpus.gutenberg.words('shakespeare-macbeth.txt')
len(macbeth)


23140

In [61]:
macbeth_filtered = [w for w in macbeth if w.lower() not in sw]
len(macbeth_filtered)

14946